third_party/window/window.go

```diff
@@ -0,0 +1,87 @@
+// source: https://github.com/jbrukh/window
+package store
+
+// An array-based moving window; a moving window
+// is a queue with a maximum size and the property
+// that when the size is reached, pushing a new
+// element into the queue causes the head to be
+// popped.
+//
+// You can optimize the amount of slice copying
+// that the MovingWindow will be doing by trading
+// off with space complexity. Namely, the underlying
+// array is allocated with a size that is the
+// multiple of the intended capacity of the queue
+// so that copying is less frequent.
+type MovingWindow struct {
+	arr  []interface{}
```

Comment:
Ah! Good point @mvdan.

On Mon, Aug 17, 2015 at 3:44 PM, Daniel Martí <notifications@github.com>
wrote:

> In third_party/window/window.go
> <https://github.com/kubernetes/heapster/pull/471#discussion_r37246106>:
>
> > +package store
> > +
> > +// An array-based moving window; a moving window
> > +// is a queue with a maximum size and the property
> > +// that when the size is reached, pushing a new
> > +// element into the queue causes the head to be
> > +// popped.
> > +//
> > +// You can optimize the amount of slice copying
> > +// that the MovingWindow will be doing by trading
> > +// off with space complexity. Namely, the underlying
> > +// array is allocated with a size that is the
> > +// multiple of the intended capacity of the queue
> > +// so that copying is less frequent.
> > +type MovingWindow struct {
> > +	arr  []interface{}
>
> I don't see why that would help. I don't think Go would allow it either,
> unless you use the unsafe package. You can't cast stuff around in Go like
> you would with void* in C.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/heapster/pull/471/files#r37246106>.
>